Source/WebCore/dom/Document.cpp

```diff
@@ -2097,6 +2099,54 @@ std::optional<BoundaryPoint> Document::caretPositionFromPoint(const LayoutPoint&
     return { { node.releaseNonNull(), offset } };
 }
 
+RefPtr<CaretPosition> Document::caretPositionFromPoint(double x, double y, CaretPositionFromPointOptions options)
+{
+    if (!hasLivingRenderTree())
+        return nullptr;
+
+    LayoutPoint clientPoint = { x, y };
+
+    LayoutPoint localPoint;
+    RefPtr node = nodeFromPoint(clientPoint, &localPoint, HitTestSource::Script);
+    if (!node)
+        return nullptr;
+
+    updateLayoutIgnorePendingStylesheets();
+
+    CheckedPtr renderer = node->renderer();
+    if (!renderer)
+        return nullptr;
+
+    if (renderer->isSkippedContentRoot())
+        return CaretPosition::create(WTFMove(node), 0);
+
+    auto position = renderer->positionForPoint(localPoint, HitTestSource::Script);
+    if (position.isNull())
+        return nullptr;
+
+    RefPtr anchorNode = position.anchorNode();
+    if (anchorNode->shadowHost()) {
+        if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(*anchorNode->shadowHost())) {
+            anchorNode = textFormControl;
+            if ((!anchorNode->isInShadowTree() && !options.shadowRoots.size()) || (anchorNode->isInShadowTree() && options.shadowRoots.contains(Ref(anchorNode->treeScope().rootNode()))))
+                return CaretPosition::create(WTFMove(anchorNode), position.offsetInContainerNode());
+        }
+    }
+
+    bool adjustPosition = false;
+    while (anchorNode->isInShadowTree() && !(options.shadowRoots.size() && options.shadowRoots.contains(Ref(anchorNode->treeScope().rootNode())))) {
```

Comment:
```suggestion
    while (anchorNode->isInShadowTree() && !(options.shadowRoots.size() && options.shadowRoots.contains(anchorNode->treeScope().rootNode()))) {
```

Reply:
This unfortunately results in a build error I am not sure how to fix:
app/webkit/Source/WebCore/dom/Document.cpp:2129:140:   required from here
/app/webkit/WebKitBuild/GTK/Release/WTF/Headers/wtf/CheckedRef.h:302:46: error: static assertion failed: Derived class should opt-in when defaulting operator==, or invalid/undefined comparison should be reworked/defined
  302 |         static_assert(defaultedOperatorEqual == DefaultedOperatorEqual::Yes, "Derived class should opt-in when defaulting operator==, or invalid/undefined comparison should be reworked/defined");
      |                       ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/app/webkit/WebKitBuild/GTK/Release/WTF/Headers/wtf/CheckedRef.h:302:46: note: ‘(WTF::DefaultedOperatorEqual::No == WTF::DefaultedOperatorEqual::Yes)’ evaluates to false
[1202/2373] Building CXX object Source/WebCore/CMakeFiles/WebCore.dir/__/__/WebCore/DerivedSources/unified-sources/UnifiedSource-be65d27a-17.cpp.o

Reply:
OK. A bit mysterious but not worth major effort I guess.